conda/core/link.py

```diff
@@ -188,6 +188,15 @@ def __init__(self, *setups):
         self.execute_executor = (DummyExecutor() if context.debug or context.execute_threads == 1
                                  else ThreadLimitedThreadPoolExecutor(context.execute_threads))
 
+    @property
+    def adds_prereleases(self) -> bool:
+        """True if a new/updated spec is a prerelease version"""
+        for stp in self.prefix_setups.values():
+            for prec in list(stp.link_precs) + list(stp.update_specs):
+                if context.is_prerelease(prec):
+                    return True
+        return False
+
```

Comment:
Is this doing a per-package check or a per-transaction check? I understand that the PEP440 specifies that if a _package_ is already using a pre-release, then it should be able to update to a pre-release too. However, this doesn't imply that the whole environment should opt-in. Is that too strict of a reading? Happy to hear otherwise, but needs to be well documented.

Reply:
This is doing a per-package check. I see nothing in the PEP 440 spec that suggests that an existing pre-release version should enable an update to a different pre-release:

> By default, dependency resolution tools SHOULD:
>
> * accept already installed pre-releases for all version specifiers
> * accept remotely available pre-releases for version specifiers where there is no final or post release that satisfies the version specifier
> * exclude all other pre-releases from consideration

So I think this implementation satisfies the spec. However, I did not look to see how pip implements this.

I think it would be reasonable to allow an installed pre-release to upgrade to a new pre-release, but only if there is no intervening regular release. For instance, if you already had `foo 1.2.3.dev2` installed, it would be ok to upgrade to `foo 1.2.3.dev4` as long as there is no `foo 1.2.4` that satisfies the original match spec. I am not sure how to implement that or whether it would be worth the trouble.